org.lflang/src/org/lflang/generator/ReactionInstanceGraph.xtend

```diff
@@ -69,8 +69,10 @@ class ReactionInstanceGraph extends DirectedGraph<ReactionInstance> {
         addNodesAndEdges(main)
             // Assign a level to each reaction. 
             // If there are cycles present in the graph, it will be detected here.
-            if (!assignLevels()) {
+            val leftoverReactions = assignLevels()
+            if (leftoverReactions !== null) {
                 main.reporter.reportError("Reactions form a cycle!");
+                System.err.println(leftoverReactions.toString());
```

Comment:
We have a much more sophisticated reporting of cycles in the validator, and I would expect this code not to be executed at all if cycles were found during validation. Is the concern that AST transformations would introduce cycles, and that's why we're reporting them here?

Reply:
It would surprise me if it would be an "expected" error when an AST transformation to introduces a cycle. In other words, if we implement AST transformations correctly, my guess is that the user should never see this message.

Reply:
Correct.  This change was prompted by the bug that this PR fixes, where cycles are introduced in AST transformations.  I've added a comment to this effect.